Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,647 @@
+                                                          NSDecimal* right,
+                                                          NSRoundingMode roundingMode);
+
+static NSCalculationError _subtraction(NSDecimal* result, const NSDecimal* left, const NSDecimal* right);
+static NSCalculationError _mantissaAddition(NSDecimalInternal* result, NSDecimalInternal* augend, unsigned short addend);
+static NSCalculationError _mantissaMultiplication(NSDecimalInternal* result, NSDecimalInternal* multiplicand, unsigned short multiplier);
+static unsigned int _obtainLength(unsigned short* mantissa, int length);
+static NSCalculationError _mantissaDivision(NSDecimalInternal* result, NSDecimalInternal* dividend, unsigned short divisor);
+static NSCalculationError _compactMantissa(NSDecimalInternal* iDecimal, NSRoundingMode roundingMode);
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCopy(NSDecimal* destination, const NSDecimal* source) {
+    if (!source || !destination) {
+        return;
+    }
+
+    destination->_isNegative = source->_isNegative;
+    destination->_exponent = source->_exponent;
+    destination->_length = source->_length;
+    destination->_isCompact = source->_isCompact;
+    destination->_reserved = source->_reserved;
+    memcpy(&destination->_mantissa, &source->_mantissa, NSDecimalMaxSize * sizeof(unsigned short));
+}
+
+/**
+ @Status Interoperable
+*/
+BOOL NSDecimalIsNotANumber(const NSDecimal* dcm) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, dcm != NULL);
+    return !dcm->_length && IS_NEGATIVE(dcm);
+}
+
+/**
+ @Status Interoperable
+*/
+NSComparisonResult NSDecimalCompare(const NSDecimal* leftOperand, const NSDecimal* rightOperand) {
+    if (leftOperand == rightOperand) {
+        return NSOrderedSame;
+    }
+
+    THROW_NS_IF_FALSE(E_INVALIDARG, (leftOperand != NULL) && (rightOperand != NULL));
+
+    // not a number
+    if (NSDecimalIsNotANumber(leftOperand) || NSDecimalIsNotANumber(rightOperand)) {
+        if (NSDecimalIsNotANumber(leftOperand) && NSDecimalIsNotANumber(rightOperand)) {
+            return NSOrderedSame;
+        }
+
+        // R > NaN
+        if (NSDecimalIsNotANumber(rightOperand)) {
+            return NSOrderedDescending;
+        }
+
+        return NSOrderedAscending;
+    }
+
+    NSDecimal left, right;
+    NSDecimalCopy(&left, leftOperand);
+    NSDecimalCopy(&right, rightOperand);
+    NSDecimalNormalize(&left, &right, NSRoundPlain);
+
+    if (!left._isNegative && right._isNegative) {
+        return NSOrderedDescending;
+    }
+
+    if (left._isNegative && !right._isNegative) {
+        return NSOrderedAscending;
+    }
+
+    // same signs
+    if (left._length < right._length) {
+        return right._isNegative ? NSOrderedDescending : NSOrderedAscending;
+    }
+
+    if (left._length > right._length) {
+        return right._isNegative ? NSOrderedAscending : NSOrderedDescending;
+    }
+
+    // all same
+    int index = right._length - 1;
+    while (index >= 0) {
```

Comment:
this should just be a for statement - no reason to go through the trouble of setting up all three clauses separately...